cmake/external/xnnpack.cmake

```diff
@@ -24,43 +24,67 @@ set(onnxruntime_EXTERNAL_LIBRARIES_XNNPACK XNNPACK pthreadpool)
 list(APPEND onnxruntime_EXTERNAL_LIBRARIES ${onnxruntime_EXTERNAL_LIBRARIES_XNNPACK})
 
 # the XNNPACK CMake setup doesn't include the WASM kernels so we have to manually set those up
-if (onnxruntime_BUILD_WEBASSEMBLY)
-  if (onnxruntime_ENABLE_WEBASSEMBLY_THREADS)
+if(onnxruntime_BUILD_WEBASSEMBLY)
+  file(READ "${XNNPACK_DIR}/BUILD.bazel" xnnpack_bazel_config)
+
+  # Replace newlines with semicolon so that it is treated as a list by CMake
+  # Also replace '[' and ']' so the bazel source lists don't get parsed as a nested list by cmake
+  string(REPLACE "\n" ";" xnnpack_bazel_config "${xnnpack_bazel_config}")
+  string(REPLACE "[" "{" xnnpack_bazel_config "${xnnpack_bazel_config}")
+  string(REPLACE "]" "}" xnnpack_bazel_config "${xnnpack_bazel_config}")
+
+  function(GetSrcListFromBazel src_list_name target_srcs)
+    set(_InSection FALSE)
+    set(bazel_srcs "")
+
+    foreach(_line ${xnnpack_bazel_config})
+      if(NOT _InSection)
+        if(_line MATCHES "^${src_list_name} = \\{")
+          set(_InSection TRUE)
+        endif()
+      else()
+        if(_line MATCHES "^\\}")
+          set(_InSection FALSE)
+        else()
+          # parse filename from quoted string with trailing comma
+          string(REPLACE "\"" "" _line "${_line}")
+          string(REPLACE "," "" _line "${_line}")
+          string(STRIP "${_line}" _line)
+
+          list(APPEND bazel_srcs "${XNNPACK_DIR}/${_line}")
+        endif()
+      endif()
+    endforeach()
+
+    set(${target_srcs} ${bazel_srcs} PARENT_SCOPE)
+  endfunction()
+
+  if(onnxruntime_ENABLE_WEBASSEMBLY_THREADS)
     target_compile_options(XNNPACK PRIVATE "-pthread")
   endif()
 
+  GetSrcListFromBazel("PROD_SCALAR_WASM_MICROKERNEL_SRCS" prod_scalar_wasm_srcs)
+  GetSrcListFromBazel("ALL_WASM_MICROKERNEL_SRCS" all_wasm_srcs)
+  GetSrcListFromBazel("WASM32_ASM_MICROKERNEL_SRCS" wasm32_asm_srcs)
+
+  message(DEBUG "prod_scalar_wasm_srcs: ${prod_scalar_wasm_srcs}\n")
+  message(DEBUG "all_wasm_srcs: ${all_wasm_srcs}\n")
+  message(DEBUG "wasm32_asm_srcs: ${wasm32_asm_srcs}\n")
+
   message("Adding WebAssembly Source Files to XNNPACK")
-  set(wasm_src_patterns "${XNNPACK_DIR}/src/init.c"
-                        "${XNNPACK_DIR}/src/params-init.c"
-                        "${XNNPACK_DIR}/src/qu8-avgpool/9*-minmax-scalar-c1.c"
```

Comment:
I honestly don't know how this worked. params-init.c is built in the params_lib target which is a dependency of the xnnpack target. If I manually included it I got an error about duplicate symbols.